### ISSUE TITLE
Some method references in Dex are missing

### ISSUE THREAD

**Issue by haeter525** [rizinorg/rizin#2416]
### Work environment

| Questions                                             | Answers                                                      |
| ----------------------------------------------------- | ------------------------------------------------------------ |
| OS/arch/bits (mandatory)                              | Ubuntu 20.04                                                 |
| File format of the file you reverse (mandatory)       | APK                                                          |
| Architecture/bits of the file (mandatory)             | Dalvik                                                       |
| `rizin -v` full output, **not truncated** (mandatory) | rizin 0.4.0-git @ linux-x86-64<br/>commit: 7b16c102654932cfae7f1d0ea3d2b909754ae9f9, build: 2022-03-16__08:32:49 |

### Expected behavior

Detect all the method references in Dex.

### Actual behavior

Some method references were missing.

In the following scenario, I looked into a method's instruction at address `0x496f6`. According to the disassembled result, it holds a reference to method `sendTextMessage`. However, commands `axl` and `axf` both missed that reference.

```
➜ rizin apk://13667fe3b0ad496a0cd157f34b7e0c991d72a4db.apk 
 -- Use 'e asm.offset=true' to show offsets in 16bit segment addressing mode.
[0x00028498]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
[x] Analyze function calls (aac)
[x] Analyze len bytes of instructions for references (aar)
[x] Check for classes
[x] Finding xrefs in noncode section with analysis.in=io.maps
[x] Analyze value pointers (aav)
[x] Value from 0x00000000 to 0x000af474 (aav)
[x] 0x00000000-0x000af474 in 0x0-0xaf474 (aav)
[x] Emulate functions to find computed references (aaef)
[x] Propagate noreturn information
[x] Use -AA or aaaa to perform additional experimental analysis.
[ERROR: sigdb path is unknown or invalid (path: /usr/local/share/rizin/sigdb)
[x] Applied 0 FLIRT signatures via sigdb
[0x00028498]> pd 1 @ 0x496f6
│           0x000496f6      invoke-virtual/range {v2..v7}, Landroid/telephony/SmsManager;->sendTextMessage(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Landroid/app/PendingIntent;Landroid/app/PendingIntent;)V
[0x00028498]> axl~496f6
[0x00028498]> axl~sendTextMessage
[0x00028498]> axf @ 0x496f6
[0x00028498]> 

```

### Steps to reproduce the behavior
- Open the [APK](https://github.com/quark-engine/apk-malware-samples/raw/master/13667fe3b0ad496a0cd157f34b7e0c991d72a4db.apk) with Rizin
- Type in the following commands.

```
aaa
pd 1 @ 0x496f6
axl~496f6
axl~sendTextMessage
axf @ 0x496f6
```

**Comment by ret2libc** [comment]
@wargio is this because `invoke-virtual/range` is used? Maybe it is not considered? I'm no expert...

**Comment by wargio** [comment]
probably is linked to the same bug as the other invoke-virtual

**Comment by wargio** [comment]
The method called at 0x000496f6 is an import (therefore not resolved) but the code that was supposed to do this was commented.
I pushed a fix to handle it and gave a quick check to verify there were no other jumps missing.

**Comment by haeter525** [comment]
Hi @wargio. Thanks for the correction. 
However, after days of digging, I wonder if it would be possible to have other solutions for this issue. Here are my thoughts.
1. It seems like Rizin supports resolving imports on ELF binaries. I tested USR/LS with commands adapted from the above and found this feature works.
2. Dex binaries often use external libraries for fundamental operations. Therefore, the analysis on Dex generally requires lots of tracking of the use of imports.
3. Other reversing tools like Radare2 and Jadx support this feature on Dex. Hence, I think it should be technically possible to do this work.

To sum up, I found this feature is practical and essential on Dex. And Rizin does have a similar one on ELF. Hence, I don't quite understand the reason not to do so. And I wonder if we could have more discussions about this and work out other solutions if possible. 

Look forward to receiving your opinion. :D

**Comment by wargio** [comment]
I do not understand how radare2 can support it when the jump itself is disabled.. https://github.com/radareorg/radare2/blob/master/libr/anal/p/anal_dalvik.c#L1072

**Comment by wargio** [comment]
i'm working to add also getting xrefs on this

**Comment by wargio** [comment]
@haeter525 please check the PR.
it now adds the reloc support and you will get the xrefs.

**Comment by haeter525** [comment]
> I do not understand how radare2 can support it when the jump itself is disabled.

Not sure the reason why Radare2 supports it, either. But the Xref to most imports does exist. Here's how I test it.

```
[0x0002a180]> axl~String.method.length~:1
sym.Landroid_support_v4_text_BidiFormatter.method.markAfter_Ljava_lang_String_Landroid_support_v4_text_TextDirectionHeuristicCompa+2 0x33996 ->      CALL -> 0x17834 sym.imp.Ljava_lang_String.method.length__I
[0x0002a180]> pd 1 @ 0x33996
│           0x00033996      6e10cb140400   invoke-virtual {v4}, Ljava/lang/String.length()I ; 0x14cb
```

>@haeter525 please check the PR.
>
>it now adds the reloc support and you will get the xrefs.

Thank for the quick patch!
However, some Xrefs are still missing with it. I doubt if it was caused by the caller and the callee both using the same address.

```
[0x10002fa08]> pd 1 @ 4295146684
            ;-- sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z:
            ;-- sym.Landroid_app_Activity_.onMenuItemSelected_ILandroid_view_MenuItem__Z:
┌ method.public.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z ();
│ bp: 0 (vars 0, args 0)
│ sp: 0 (vars 0, args 0)
│ rg: 0 (vars 0, args 0)
│           0x10002bcbc      invoke-super {v1, v2, v3}, Landroid/app/Activity;->onMenuItemSelected(ILandroid/view/MenuItem;)Z ; boolean onMenuItemSelected(int, android.view.MenuItem)
[0x10002fa08]> axl~onMenuItemSelected
[0x10002fa08]> axf @ 4205146684
[0x10002fa08]> 
```

```
[0x10002fa08]> pd 1 @ 4295146944
            ;-- sym.Landroid_support_v4_app_FragmentActivity_.onPostResume__V:
            ;-- sym.Landroid_app_Activity_.onPostResume__V:
┌ method.protected.Landroid_support_v4_app_FragmentActivity_.onPostResume__V ();
│ bp: 0 (vars 0, args 0)
│ sp: 0 (vars 0, args 0)
│ rg: 0 (vars 0, args 0)
│           0x10002bdc0      invoke-super {v2}, Landroid/app/Activity;->onPostResume()V ; void onPostResume()
[0x10002fa08]> axl~onPostResume
sym.Landroid_support_v4_app_FragmentActivity_.onPostResume__V+12 0x10002bdcc ->      CALL -> 0x8000000280 sym.Landroid_os_Handler_.removeMessages_I_V
sym.Landroid_support_v4_app_FragmentActivity_.onPostResume__V+18 0x10002bdd2 ->      CALL -> 0x10002bee8 method.protected.Landroid_support_v4_app_FragmentActivity_.onResumeFragments__V
sym.Landroid_support_v4_app_FragmentActivity_.onPostResume__V+28 0x10002bddc ->      CALL -> 0x10004ce2c method.public.Landroid_support_v4_app_FragmentManagerImpl_.execPendingActions__Z
[0x10002fa08]> axf @ 4295146944
[0x10002fa08]> 
```

**Comment by wargio** [comment]
Ok i have looked at it and seems that the issue was due imports also being in a structure in which i never thought it would be due the super calls

**Comment by wargio** [comment]
This is the new output from the PR above
```
[0x100028498]> 
[0x100028498]> pd 1 @ 4295146684
            ;-- sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z:
┌ method.public.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z ();
│ bp: 0 (vars 0, args 0)
│ sp: 0 (vars 0, args 0)
│ rg: 0 (vars 0, args 0)
│           0x10002bcbc      invoke-super {v1, v2, v3}, Landroid/app/Activity;->onMenuItemSelected(ILandroid/view/MenuItem;)Z ; sym.Landroid_app_Activity_.onMenuItemSelected_ILandroid_view_MenuItem__Z ; boolean onMenuItemSelected(int, android.view.MenuItem)
[0x100028498]> axl~onMenuItemSelected
method.public.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z 0x10002bcbc ->      CALL -> 0x8000000948 sym.Landroid_app_Activity_.onMenuItemSelected_ILandroid_view_MenuItem__Z
sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+24 0x10002bcd4 ->      CODE -> 0x10002bcca sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+14
sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+30 0x10002bcda ->      CALL -> 0x10004c758 method.public.Landroid_support_v4_app_FragmentManagerImpl_.dispatchOptionsItemSelected_Landroid_view_MenuItem__Z
sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+38 0x10002bce2 ->      CODE -> 0x10002bcca sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+14
sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+44 0x10002bce8 ->      CALL -> 0x10004c58c method.public.Landroid_support_v4_app_FragmentManagerImpl_.dispatchContextItemSelected_Landroid_view_MenuItem__Z
sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+52 0x10002bcf0 ->      CODE -> 0x10002bcca sym.Landroid_support_v4_app_FragmentActivity_.onMenuItemSelected_ILandroid_view_MenuItem__Z+14
[0x100028498]> axf @ 4295146684
C 0x8000000948 sym.Landroid_app_Activity_.onMenuItemSelected_ILandroid_view_MenuItem__Z
[0x100028498]> 
```

**Comment by wargio** [comment]
@haeter525 please can you confirm there are no more missing xrefs?

**Comment by haeter525** [comment]
Hi @wargio. Most xrefs work fine. Thanks.
But there are four ones still not resolved in this sample. It seems Rizin didn't check parent classes for the implementation of the callees. And I believe they are the last four missing xrefs.

Addresses: `4295335392`, `4295337516`,  `4295337568`, and `4295337616`.

```
[0x100028498]> pd 1 @ 4295335392
            ;-- sym.Landroid_support_v4_content_CursorLoader_.onReset__V:
┌ method.protected.Landroid_support_v4_content_CursorLoader_.onReset__V ();
│ bp: 0 (vars 0, args 0)
│ sp: 0 (vars 0, args 0)
│ rg: 0 (vars 0, args 0)
│           0x100059de0      invoke-super {v1}, Landroid/support/v4/content/AsyncTaskLoader;->onReset()V ; void onReset()
[0x100028498]> axf @ 4295335392
[0x100028498]> 
[0x100028498]> icm~AsyncTaskLoader~onReset
[0x100028498]> ic~AsyncTaskLoader:1
0x0001ae54 0x100051e48  0x100059d8a  Landroid/support/v4/content/AsyncTaskLoader;                                                                     Landroid/support/v4/content/Loader;
[0x100028498]> icm~Loader~onReset:0
0x100030074  600   Landroid/support/v4/content/Loader;                                                                              r          onReset()V
[0x100028498]> 
```

**Comment by wargio** [comment]
@haeter525 also that bug should be fixed. can you check again?

**Comment by haeter525** [comment]
@wargio That bug was fixed. And I will test more samples to check if any other types of xrefs are missing. Thank you for all the help! :smile:

**Comment by haeter525** [comment]
Hi @wargio. Unfortunately, I also found xrefs missing in [this sample](https://github.com/quark-engine/apk-malware-samples/raw/master/14d9f1a92dd984d6040cc41ed06e273e.apk). After some tests, I think those missing xrefs can be categorized into two groups.

1. Most of them point to methods with large sizes. And printing the bytecodes of those methods causes Rizin to show an error. Thus, I doubt there might be a parsing issue with those methods.

   Address: `0x10002ae20`, `0x100013848`, `0x100025eb4`, `0x1000194be`, and `0x100020b9c`.

```
[0x100013ca0]> pd 1 @ 0x10002ae20
│           0x10002ae20      invoke-interface/range {v15..v16}, Ljava/util/List;->add(Ljava/lang/Object;)Z
[0x100013ca0]> is~java.util.List~add:0
960  0x00003355 0x100003355  LOCAL  METH  1207961736  java.util.List                                                       boolean add(Object)
[0x100013ca0]> pdf @ 0x100003355
p: Cannot find function at 0x100003355
[0x100013ca0]> 
```

```
[0x100013ca0]> pd 1 @ 0x100013848
│           0x100013848      invoke-direct {v0}, Landroid/content/BroadcastReceiver;-><init>()V
[0x100013ca0]> is~android.content.BroadcastReceiver~init
15   0x00003304 0x100003304  LOCAL  METH  558698      android.content.BroadcastReceiver                                    void <init>()
[0x100013ca0]> pdf @ 0x100003304
p: Cannot find function at 0x100003304
[0x100013ca0]> 
```

2. The rest of them point to methods with regular sizes and printable bytecodes. Also, those methods are not imports. However,  somehow Rizin didn't resolve them.

   Address: `0x100018888`, `0x10001fb5a`, `0x100027364`, and `0x10002e704`.

```
[0x100013ca0]> pd 1 @ 0x100018888
│           0x100018888      invoke-virtual {v6}, Lcom/google/progress/ContactsCollecter;->getContactList()Ljava/lang/String;
[0x100013ca0]> icm~ContactsCollecter~getContactList
0x10001ada4  287   Lcom/google/progress/ContactsCollecter;                                p     getContactList()Ljava/lang/String;
[0x100013ca0]> 
```

```
[0x100013ca0]> pd 1 @ 0x10001fb5a
│           0x10001fb5a      invoke-virtual {v4}, Lorg/apache/commons/net/ftp/FTP;->getControlEncoding()Ljava/lang/String;
[0x100013ca0]> icm~Lorg/apache/commons/net/ftp/FTP~getControlEncoding
0x10001fe38  541   Lorg/apache/commons/net/ftp/FTP;                                       p     getControlEncoding()Ljava/lang/String;
[0x100013ca0]>
```

**Comment by wargio** [comment]
regarding those xrefs, i do not think is a mistake that they do not have valid addresses 
![image](https://user-images.githubusercontent.com/561184/161974355-6ec7f1ee-44dd-452a-8744-79770c5eef8f.png)
Tbh, i do not know what to do in this case, but i can check why those numbers are there

**Comment by haeter525** [comment]
Hi @wargio. Thanks for the reply. I'm still trying to find more clues for this case.

Currently, I found the information provided by command `icm` is wrong. For the callees in the first group, it gives library names different from those in command `is`. Moreover, according to Jadx, those methods with such library names didn't exist in this sample.

```
[0x10001122c]> is~0x100003355
960  0x00003355 0x100003355  LOCAL  METH  1207961736 java.util.List                                                       boolean add(Object)
[0x10001122c]> icm~0x100003355
0x100003355  1663  Lorg/apache/commons/net/util/SubnetUtils$1;                                  add(Ljava/lang/Object;)Z
[0x10001122c]> 
```
![Screenshot from 2022-04-08 10-33-46](https://user-images.githubusercontent.com/3461569/162355450-0b84491e-dec0-4829-9740-48088c1dde7f.png)

Since a library name is part of the identifier for a method, I wonder if it might be a possible cause for the missing xrefs. However, I may need more study on the codebase to confirm that.

**Comment by wargio** [comment]
i have found the issue, now i have to find the exact cause.

**Comment by wargio** [comment]
i have found the cause. the issue is due a missing of a small check which is related to the data section.
i will add the command `dexs` to dump the dex structure which outputs like `dexdump`

**Comment by wargio** [comment]
https://github.com/rizinorg/rizin/pull/2535 @haeter525 can you check again? i have verified the whole dex by hand with dexdump, looks the same to me.

**Comment by haeter525** [comment]
@wargio #2535 fixed the above issue. The xrefs are all resolved. Greatly appreciate your help!

**Comment by wargio** [comment]
no, thank you for finding the bug and validating the code :)
